scripts/rebuild-filterlist.mjs

```diff
@@ -3,28 +3,171 @@
 
 const easylistRevision = fs.readFileSync(path.join(rulesDir, 'filterlists', 'easylist_revision.txt'), 'utf-8');
 
-// TODO: consider using python-abp (flrender) to generate filterlist properly
-const filterlistContent = `
-[Adblock Plus 2.0]
+const trancoCSV = fs.readFileSync(path.join(dataDir, 'top-1m.csv'), 'utf-8');
+const MAX_DOMAIN_RANK = 100000;
+
+let domainMap
+
+async function processFilterList (listFileName) {
+    console.log(`Processing ${listFileName}`);
+    const data = fs.readFileSync(path.join(rulesDir, 'filterlists', listFileName), 'utf-8');
+    const lines = data.split('\n');
+    const filterlistJSON = {"rules": {}};
+
+    // first remove irrelevant rules
+    const filteredLines = lines.filter(line => !line.startsWith('!') &&
+                                               !line.startsWith('##') &&
+                                               !line.startsWith('||') &&
+                                               !line.includes(':remove') &&
+                                               !line.includes(':upward') &&
+                                               !line.includes('redirect-rule'));
+    // dump rules into json structure for parsing
+    for (const line in filteredLines) {
+        const rule = filteredLines[line];
+        const splitRule = rule.split('##');
+        const target = splitRule[0];
+        const action = splitRule[1];
+
+        if (!target || !action) {
+            continue;
+        }
+
+        if (!filterlistJSON.rules[target]) {
+            filterlistJSON.rules[target] = [action];
+        } else {
+            filterlistJSON.rules[target].push(action);
+        }
+    }
+
+    // remove any targets that contain a +js rule. we do it this way because
+    // when a target has a :js rule, we must also remove any css rules
+    for (const item in filterlistJSON.rules) {
+        const containsJS = filterlistJSON.rules[item].some(filter => filter.includes('+js'));
+        if (containsJS) {
+            delete filterlistJSON.rules[item];
+        }
+    }
+
+    // remove longtail domain rules using tranco data
+    for (const item in filterlistJSON.rules) {
+        let domains;
+
+        if (item.includes(',')) {
+            domains = item.split(',');
+        } else {
+            domains = [item];
+        }
+
+        const filteredDomains = domains.filter(domain => {
+            const tlDomain = getDomain(domain);
+            return isPopularDomain(tlDomain);
+        });
+
+        // if no domains meet cutoff, remove rule entirely. if one or more
+        // domains meet cutoff, remove those that don't
+        if (filteredDomains.length === 0) {
+            delete filterlistJSON.rules[item];
+        } else if (filteredDomains.length !== domains.length) {
+            const filteredDomainString = filteredDomains.join(',');
+            filterlistJSON.rules[filteredDomainString] = filterlistJSON.rules[item];
+            delete filterlistJSON.rules[item];
```

Comment:
Removing the item while iterating through the object is not a problem in JS, right? Just double checking

Reply:
Yep, shouldn't be a problem.